### PR TITLE
Feature/tagger

### LINKED ISSUE #458: Tagging with propagation (Server)
Tagging should be possible with a propagation request on a given field and with a given filter (taken from a bookmark).

Furthermore in order to be able to replay the tagging, the tagging service should now be independant from the server and make use of 2 kafkas topics:  
- a `tagref_log` topic to hold the tagging order (with or without propagation)
- a `execute_tags` topic to hold the actual tagging requests to the relevant hits

### PR DESCRIPTION
**Pull request by alainbodiguel**
Fix #458

### CODE REVIEW

File: arlas-tagger/src/main/java/io/arlas/server/rest/tag/TagStatusRESTService.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    protected static Logger LOGGER = LoggerFactory.getLogger(TagStatusRESTService.class);
+    public static final String UTF8JSON = MediaType.APPLICATION_JSON + ";charset=utf-8";
+    private TaggingStatus status;
+
+    public TagStatusRESTService() {
+        this.status = TaggingStatus.getInstance();
+    }
+
+    @Timed
+    @Path("/{collection}/_tagging")
```
Comment (sylvaingaudan) [inline arlas-tagger/src/main/java/io/arlas/server/rest/tag/TagStatusRESTService.java]: we use verbs like _search or _aggregate or _compute

File: arlas-tagger/src/main/java/io/arlas/server/service/KafkaConsumerRunner.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    public abstract void processRecords(ConsumerRecords<String, String> records);
+
+    @Override
+    public void run() {
+        try {
+            LOGGER.info("Starting consumer of topic " + topic);
+            consumer = TagKafkaConsumer.build(configuration, topic, consumerGroupId);
+            while (true) {
+
+                ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(100l));
```
Comment (sylvaingaudan) [inline arlas-tagger/src/main/java/io/arlas/server/service/KafkaConsumerRunner.java]: => configurable with default value

File: arlas-tagger/src/main/java/io/arlas/server/service/TagRefService.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        if (records.count() > 0) {
+            LOGGER.debug("Kafka polling returned " + records.count());
+        }
+
+        for (ConsumerRecord<String, String> record : records) {
+
+            try {
+                final TagRefRequest tagRequest = MAPPER.readValue(record.value(), TagRefRequest.class);
+                if (tagRequest.propagation == null) {
+                    LOGGER.info("No propagation requested: " + record.value());
```
Comment (sylvaingaudan) [inline arlas-tagger/src/main/java/io/arlas/server/service/TagRefService.java]: anything that is printed at every message should be debug and not info

File: arlas-tagger/src/main/java/io/arlas/server/service/TagRefService.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                    tagKafkaProducer.sendToExecuteTags(tagRequest);
+                } else {
+                    LOGGER.info("Propagation requested: " + record.value());
```
Comment (sylvaingaudan) [inline arlas-tagger/src/main/java/io/arlas/server/service/TagRefService.java]: anything that is printed at every message should be debug and not info

File: arlas-tagger/src/main/java/io/arlas/server/service/TagRefService.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+                        Search search = new Search();
+                        search.filter = filter;
+                        tagKafkaProducer.sendToExecuteTags(TagRefRequest.fromTagRefRequest(tagRequest, search, (float)(i+1)/nbResult));
+                        taggingStatus.updateStatus(tagRequest.id, updateResponse);
+
+                    }
+                }
+            } catch (IOException e) {
+                LOGGER.warn("Could not parse record " + record.value());
```
Comment (sylvaingaudan) [inline arlas-tagger/src/main/java/io/arlas/server/service/TagRefService.java]: it is an error, not a warning

Comment (alainbodiguel) [reply arlas-tagger/src/main/java/io/arlas/server/service/TagRefService.java]: As agreed, we leave as a warning.

File: arlas-tagger/src/main/java/io/arlas/server/app/ArlasTagger.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        bootstrap.registerMetrics();
+        bootstrap.setConfigurationSourceProvider(new SubstitutingSourceProvider(
+                bootstrap.getConfigurationSourceProvider(), new EnvironmentVariableSubstitutor(false)));
+        bootstrap.addBundle(new SwaggerBundle<ArlasServerConfiguration>() {
+            @Override
+            protected SwaggerBundleConfiguration getSwaggerBundleConfiguration(ArlasServerConfiguration configuration) {
+                return configuration.swaggerBundleConfiguration;
+            }
+        });
+        bootstrap.addBundle(new ZipkinBundle<ArlasServerConfiguration>(getName()) {
```
[the remark this answers is not in the document]
Comment (alainbodiguel) [reply arlas-tagger/src/main/java/io/arlas/server/app/ArlasTagger.java]: Change committed.

File: arlas-tagger/src/main/java/io/arlas/server/app/ArlasTagger.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        Settings settings = settingsBuilder.build();
+
+        PreBuiltTransportClient transportClient = new PreBuiltTransportClient(settings);
+        for(Pair<String,Integer> node : configuration.getElasticNodes()) {
+            transportClient.addTransportAddress(new TransportAddress(InetAddress.getByName(node.getLeft()),
+                    node.getRight()));
+        }
+        Client client = transportClient;
+
+        if (configuration.zipkinConfiguration != null) {
```
[the remark this answers is not in the document]
Comment (alainbodiguel) [reply arlas-tagger/src/main/java/io/arlas/server/app/ArlasTagger.java]: Change committed.

File: arlas-tagger/src/main/java/io/arlas/server/app/ArlasTagger.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        environment.jersey().register(InsensitiveCaseFilter.class);
+
+        //healthchecks
+        environment.healthChecks().register("elasticsearch", new ElasticsearchHealthCheck(client));
+
+        //cors
+        if (configuration.arlascorsenabled) {
+            configureCors(environment);
+        }
+        NodesInfoRequest nodesInfoRequest = new NodesInfoRequest();
```
[the remark this answers is not in the document]
Comment (alainbodiguel) [reply arlas-tagger/src/main/java/io/arlas/server/app/ArlasTagger.java]: Change committed.

File: arlas-tagger/src/main/java/io/arlas/server/model/TaggingStatus.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import io.arlas.server.model.response.UpdateResponse;
+import io.arlas.server.util.SelfExpiringHashMap;
+import io.arlas.server.util.SelfExpiringMap;
+
+import java.util.Optional;
+
+public class TaggingStatus {
+    private SelfExpiringMap<String, UpdateResponse> statusMap;
+    private TaggingStatus() {
+        statusMap = new SelfExpiringHashMap<>(3600000l);
```
[the remark this answers is not in the document]
Comment (alainbodiguel) [reply arlas-tagger/src/main/java/io/arlas/server/model/TaggingStatus.java]: Change committed.

File: arlas-tagger/src/main/java/io/arlas/server/rest/tag/TagStatusRESTService.java (hunk 1 above)
Comment (alainbodiguel) [reply arlas-tagger/src/main/java/io/arlas/server/rest/tag/TagStatusRESTService.java]: Change committed.

File: arlas-tagger/src/main/java/io/arlas/server/service/KafkaConsumerRunner.java (hunk 2 above)
Comment (alainbodiguel) [reply arlas-tagger/src/main/java/io/arlas/server/service/KafkaConsumerRunner.java]: Change committed.

File: arlas-tagger/src/main/java/io/arlas/server/service/TagRefService.java (hunk 4 above)
Comment (alainbodiguel) [reply arlas-tagger/src/main/java/io/arlas/server/service/TagRefService.java]: Change committed.

File: arlas-tagger/pom.xml [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    <properties>
+        <kafka-connect.version>2.1.1</kafka-connect.version>
+        <kafka.version>2.1.1</kafka.version>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.arlas</groupId>
+            <artifactId>arlas-opensearch</artifactId>
+            <version>10.6.2-SNAPSHOT</version>
```
Comment (MohamedHamouGisaia) [inline arlas-tagger/pom.xml]: arlas-opensearch is not needed

**Comment by MohamedHamouGisaia** [comment]
`docs/arlas-server-configuration.md` contains all the configuration variable => for tagging configuration
Otherwise, LGTM

**Pull request merged**